requirements.txt

```diff
@@ -1,3 +1,4 @@
 # mail_inline_css
 premailer
 extract-msg
+freezegun==0.3.11
```

Comment:
With latest freezegun it just fails with some voodoo errors :) And since it's first time I'm using freezegun, I have no idea what those were about. @pedrobaeza are you going to complain? :)

Reply:
Well, firstly I would complain why you need an extra library for such basic feature...

Reply:
@simahawk requested to use it for substituting the "current time" in tests, previously it was done using context

Reply:
Yeah, and that's why I don't like to use a library only for testing and for this specific thing, just as it happened with one of his modules. Please revert to other mechanism. I remember that Odoo has added a mechanism to manipulate dates obtained by `fields.Date/Datetime.today/now`, but I would say this is for v13.

Reply:
@alexey-pelykh was a joke because Pedro complained about adding the same dep in another repo. I'm totally in favor of it because it's pointless to replicate the same patch all around.
And many times the patches do not work like freezegun because it does it right and everywhere, plus it has some nice features.
IMO it should be included in our tests toolset and recommended as the right way to tests w/ dates. Then we can remove all the hacks that have been made (eg: ctx key for tests only).

Reply:
I don't like a library only for tests...

Reply:
But that library is only for tests, and QUnit is only for tests. I do agree that including library only for tests in every repo is a bit silly, yet injecting code in _production_ to achieve the same is as silly to me. So in this case I'd vote for actually adding this library to the test tools and to the guidelines "how to simulate datetime"

Reply:
Yet in any case, please make up your minds :) Changing back and forth is a bit depressing

Reply:
Why don't you use `context_today` https://github.com/odoo/odoo/blob/939bca38212d5eba830bbf70d23f500ace578b1a/odoo/fields.py#L1666 and simply pass the call with timestamp in a context provided in the test? No need of voodoo or library.

Reply:
That's how it was done before, almost. A ctx variable defaulting to `now()` (and I need datetime not date). As I've said, there are opposite requests to change the PR: one to use freezegun and don't workaround with context, other one to workaround with context and not to use library. Once @pedrobaeza and @simahawk are in agreement how to proceed, I'd gladly adopt the OCA way :)

Reply:
I have it totally clear: this is not a workaround, it's the way Odoo intends to force other timestamps than now, so we should use it.

Reply:
Technically, there's no method that would return UTC now or timestamp if provided in `fields.Datetime`, so I'll have to use:
```
self.env.context.get('mail_activity_reminder_now', fields.Datetime.now())
```
which is kind-a workaround, and is not as universal as replacing what `now` returns. Yet definitely is more "in-place" than using library. @simahawk do you agree in a non-blocking way with Pedro?

Reply:
What I suggest is the following in your code (inexact code):
```python
fields.Datetime.context_timestamp(timestamp=self.env.context('override_timestamp'))
```
and in tests, pass that context:
```python
model.with_context(override_timestamp=x).call_cron
```

Reply:
I got that, it was almost implemented like that, but context_timestamp is useless for me, so it's similar to ```self.env.context.get('mail_activity_reminder_now', fields.Datetime.now())```

Reply:
OK, that is perfectly fine:

- no extra libraries
- no overhead
- almost the same code as adding the freezegun stuff
- no extra knowledge needed

Reply:
IMO the point is that is silly to have code that is needed only for tests.
And in EVERY place where you need the date to match yours, you'll have to take care of it.

@pedrobaeza how is this different from using openupgradelib for JUST running an upgrade step that can be done w/ pure odoo code? You do it because it's handy and it avoids copy/pasting things all around. I understand it. Here is the same.

"no extra knowledge needed" is wrong: you must know that to test that module you must pass that ctx key and every time someone will read the code he/she will try to figure out why you need that thing. Furthermore, if this is the pattern you establish, you'll have ppl adapting it to their use cases which will make it every time a bit different to understand.

If you use freezegun you don't care. You test always in the same way.

Moreover, tests are more clear: if the test is decorated w/ `@freeze_time(...)` you know since the beginning at glance what you are going to do.

Also, maybe not the case here, but if someone extends the module and does something related to dates he/she will have to replicate again the same trick just because you make it functional.

My $0.02 :)

@sbidoul @lmignon @sebastienbeau @guewen can you share your thoughts regarding this?
Would be nice to solve this once for all :pray:

Reply:
I quickly skimmed through this thread. So the problem is about adding a python lib dependency that is necessary for testing only?

In an ideal world Odoo should have a manifest key for test dependencies. Or `python-tests` in `external_dependencies`. Or something like that. Did someone request that upstream yet?

Now MQT recently gained a [new way](https://github.com/OCA/maintainer-quality-tools/blob/9423a4623e10ee4e91decb38c7eb546768fb555b/travis/travis_install_nightly#L119-L134) to discover dependencies automatically, and incidentally has a mechanism to declare test dependencies at the repo level. (note this will ignore `requirements.txt` and `oca_dependencies.txt`)

So in this case you can add `freezegun` to `test-requirements.txt` and set `MQT_DEP=PIP` in `.travis.yml`.

Reply:
Personally, I find freezegun very useful to manage the dates into our unittests. I know it's an extra dependency just required for tests but it's not the first one and IMO it add a plus value. I am not in favour of having to write specific code into the business logic used only by the tests.  Even if the intention is good, this type of code could be interpreted as part of the API of the module. We are going to see people using this code in their development when we must ban the use of the context as a means of passing arguments. For all of this, I'm in favour of using freezegun.

Reply:
OK, proceed that way

Reply:
So, that way being...?

Reply:
The library one